src/bluesky/simulators.py

```diff
@@ -102,5 +117,285 @@ async def check_limits_async(plan):
+    >>> messages = sim.simulate_plan(my_plan())
+    >>>
+    """
+
+    GROUP_ANY = "any"
+
+    def __init__(self):
+        self.message_handlers = []
+        self.callbacks = {}
+        self.next_callback_token = 0
+        self.return_value: Any
+
+    def add_handler_for_callback_subscribes(self):
+        """Add a handler that registers all the callbacks from subscribe messages so we can call them later.
+
+        You will need to call this as one of the first things if you wish to fire callbacks from the simulator.
+        """
+        self.message_handlers.append(
+            _MessageHandler(
+                lambda msg: msg.command == "subscribe",
+                lambda msg: self._add_callback(msg.args),
+            )
+        )
+
+    def add_handler(
+        self,
+        commands: Union[str, Sequence[str]],
+        handler: Callable[[Msg], object],
+        msg_filter: Optional[Union[str, Callable[[Msg], bool]]] = None,
+        index: Union[int, Literal["end"]] = 0,
+    ):
+        """Add the specified handler for a particular message.
+
+        Parameters
+        ----------
+        commands
+            The command name for the message as defined in bluesky Message Protocol, or a sequence if
+            more than one matches.
+        handler
+            A lambda that accepts a Msg and returns an object; the object is sent to the current yield
+            statement in the generator, and is used when reading values from devices, the structure of the
+            object depends on device hinting.
+        msg_filter
+            This can either be a string corresponding to the Msg obj.name property to match, or it can
+            be a predicate on the Msg which returns a bool. The default is None which will match all
+            messages.
+        index
+            An optional integer indicating where to insert the handler, the default is 0 which
+            prepends it so that newer handlers override older ones. Specify END to append the handler.
+
+        """
+        if isinstance(commands, str):
+            commands = [commands]
+
+        self.message_handlers.insert(
+            index if index != END else len(self.message_handlers),
+            _MessageHandler(
+                lambda msg: msg.command in commands
+                and (
+                    msg_filter is None
+                    or (callable(msg_filter) and msg_filter(msg))
+                    or (msg.obj and msg.obj.name == msg_filter)
+                ),
+                handler,
+            ),
+        )
+
+    def add_read_handler_for(self, obj_name: str, value: Any, dict_key: str = "values"):
+        """
+        Convenience method to register a handler to return a result from a 'read' command.
+
+        Parameters
+        ----------
+        obj_name
+            Name of the object to intercept reads for
+        value
+            The value to return. If value is a dict then dict_key is ignored and the result of the read
```

Comment:
Ignoring inputs of one parameter based on the value of another is generally a poor pattern.  Could this be a good case for `**kwargs` where if the value is a dict, it is anything else, gets a timestamp automatically appended?

```python
sim.add_read_handler("pin_tip_detection-triggered_tip", a=(8, 5), b=5, c={'value': 'bob', 'timestamp': 0})
```

which would return

```python
{
 'a': {'value': (8, 5}, 'timestamp': XXX},
 'b': {'value': 5, 'timestamp': XXX},
 'c': {'value': 'bob', 'timestamp': 0}
}
```

on `'read'`?

Reply:
I've added `add_read_handler_for_multiple` in order to separate the two cases, likewise `add_callback_handler_for` now has `add_callback_handler_for_multiple`